game-core/src/main/java/games/strategy/triplea/ai/pro/util/ProTerritoryValueUtils.java

```diff
@@ -429,18 +431,41 @@ private static double findWaterValue(
     return capitalOrFactoryValue / 100 + nearbyLandValue / 10;
   }
 
+  // 1700 oldT=4906 newT=227 newT2=166 newT3=152
+  @SuppressWarnings("ReferenceEquality")
   private static Set<Territory> findNearbyEnemyCapitalsAndFactories(
-      final Territory t, final Map<Territory, Double> enemyCapitalsAndFactoriesMap) {
-
-    Set<Territory> nearbyEnemyCapitalsAndFactories = new HashSet<>();
-    for (int i = MIN_FACTORY_CHECK_DISTANCE; i <= MAX_FACTORY_CHECK_DISTANCE; i++) {
-      nearbyEnemyCapitalsAndFactories = t.getData().getMap().getNeighbors(t, i);
-      nearbyEnemyCapitalsAndFactories.retainAll(enemyCapitalsAndFactoriesMap.keySet());
-      if (!nearbyEnemyCapitalsAndFactories.isEmpty()) {
-        break;
+      final Territory t, final Set<Territory> enemyCapitalsAndFactories) {
+    final GameMap map = t.getData().getMap();
+
+    int distance = 0;
+    Territory nextDistanceAfter = t;
+
+    final var found = new HashSet<Territory>();
+    final var visited = new HashSet<Territory>(List.of(t));
+    final var queue = new ArrayDeque<Territory>(List.of(t));
+    while (!queue.isEmpty()) {
+      final Territory node = queue.removeFirst();
+      for (final Territory t2 : map.getNeighbors(node)) {
+        if (visited.add(t2)) {
+          queue.add(t2);
+          if (enemyCapitalsAndFactories.contains(t2)) {
+            found.add(t2);
+            if (found.size() == enemyCapitalsAndFactories.size()) {
+              break;
+            }
+          }
+        }
       }
-    }
 
-    return nearbyEnemyCapitalsAndFactories;
+      if (node == nextDistanceAfter) {
+        distance++;
```

Comment:
I undertand how this distance is counted, pretty clever way to do it efficiently, however a comment explaining it would be nice in case we ever look at this code again (which will likely happen)

Reply:
Added a comment, thanks!